### DISCUSSION TITLE
TMPDIR not writeable; how to set TMPDIR with .env?

### DISCUSSION THREAD

**Discussion by davised** [casey/just#1269] [Q&A] [answered]
Hi All,

I have a situation on my compute cluster where TMPDIR is set to a value that works on most nodes but not all. I have to manually change the TMPDIR variable on a few select machines because the standard location is not available.

I was hoping that I would be able to use a `.env` file and use `set dotenv-load` in my justfile to circumvent this issue, but I'm still getting a 'permission denied' error as `just` is attempting to write to the original TMPDIR location. [Edit] - It seems like variables set outside of the `.env` file take precedence over the variables set in the `.env` file, is this true? [/Edit]

Is there any workaround other than exporting a new TMPDIR before running `just`? I can run `just -l` with no issue but it seems that recipes that start with a `#!/usr/bin/env bash` have an issue - I'm assuming because `just` is writing a shell script to the TMPDIR to run.

Thanks,

Ed

**Comment by davised** [comment]
Self, you can do this:

```
set export

TMPDIR := "/tmp"
```

And it will set the TMPDIR variable for you.

It doesn't seem as though one can use a .env file for this purpose.

**Reply by davised** [reply]
Assuming you don't use a `#!/usr/bin/env bash` recipe. You need to use semi-colon separated commands.

**Comment by casey** [comment, marked as answer]
I love that `just` has made it onto a compute cluster!

The issue is that `.env` files change environment variables for child processes of `just`, e.g. recipe lines and shebang recipes, but not for `just` itself. So if you set `TMPDIR` in a `.env` file or with `export TMPDIR := "/tmp"`, just will still use the old tempdir.

This would be very easy to configure though, we could add a setting specifically for setting the tempdir that just uses, like so:

```
set tempdir := "/tmp"
```

Would that work for your use-case?

**Reply by davised** [reply]
That would be great!

And yes, as both a user of the compute cluster and someone who manages installs, I've been loving `just`! I use it for processing data, deploying updates to static sites using hugo, and testing out software installs.

**Reply by casey** [reply]
Nice, that's awesome to hear! I created #1300 to track this.

**Reply by casey** [reply]
@davised This has been implemented in #1369 by @dmatos2012! Currently on master but it should make it into a release soon. Give it a try!

**Reply by davised** [reply]
I gave it a spin. I'll document below.

**Reply by davised** [reply]
Here is my file:
```make
#!/usr/bin/env just
# set tempdir := "/tmp"
# set export
# TMPDIR := "/tmp"

_default:
  @just -l

deploy:
  git push; dest=/data/www/software/html/updates; hugo; cp -r public/* $dest

deploy-sh:
  #!/usr/bin/env bash
  git push
  dest=/data/www/software/html/updates
  hugo
  cp -r public/* $dest
```
I have two recipes that do the same thing to document running as a recipe without the shebang and with the shebang.

**Reply by davised** [reply]
Here are the errors if no tmpdirs are set and the TMPDIR env var is invalid:
```shell
$ just deploy
git push; dest=/data/www/software/html/updates; hugo; cp -r public/* $dest
Everything up-to-date
Total in 2 ms
panic: mkdir /data/hugo_cache: permission denied

goroutine 1 [running]:
github.com/spf13/afero.GetTempDir(0x29170e0, 0x37aa300, 0x247a3dc, 0xa, 0xc000adf180, 0xc000b3e648)
	/home/cgrb/cgrbinst/go/pkg/mod/github.com/spf13/afero@v1.6.0/util.go:128 +0x2b7
github.com/gohugoio/hugo/helpers.GetTempDir(...)
	/local/downloads/hugo-extended/hugo-0.87.0/helpers/path.go:441
github.com/gohugoio/hugo/helpers.GetCacheDir(0x29170e0, 0x37aa300, 0x2917ab0, 0xc0002ac870, 0x2a, 0x28b3000, 0x1, 0x9)
	/local/downloads/hugo-extended/hugo-0.87.0/helpers/path.go:405 +0x19a
github.com/gohugoio/hugo/cache/filecache.resolveDirPlaceholder(0x29170e0, 0x37aa300, 0x2917ab0, 0xc0002ac870, 0x24b5387, 0x9, 0xc0001e44c0, 0x2, 0x2, 0x0, ...)
	/local/downloads/hugo-extended/hugo-0.87.0/cache/filecache/filecache_config.go:228 +0x205
github.com/gohugoio/hugo/cache/filecache.DecodeConfig(0x29170e0, 0x37aa300, 0x2917ab0, 0xc0002ac870, 0x0, 0x31, 0xc0004c4da0)
	/local/downloads/hugo-extended/hugo-0.87.0/cache/filecache/filecache_config.go:170 +0xdd4
github.com/gohugoio/hugo/hugolib.configLoader.collectModules(0x2917ab0, 0xc0002ac870, 0x29170e0, 0x37aa300, 0x291a138, 0xc000a9bc20, 0x0, 0x0, 0xc0000421b4, 0x2a, ...)
	/local/downloads/hugo-extended/hugo-0.87.0/hugolib/config.go:372 +0x165
github.com/gohugoio/hugo/hugolib.LoadConfig(0x29170e0, 0x37aa300, 0x291a138, 0xc000a9bc20, 0x0, 0x0, 0xc0000421b4,0x2a, 0xc0000421b4, 0x2a, ...)
	/local/downloads/hugo-extended/hugo-0.87.0/hugolib/config.go:154 +0xa30
github.com/gohugoio/hugo/commands.(*commandeer).loadConfig(0xc0001b9340, 0xc0001b0001, 0x28e3c20, 0xc00034c210)
	/local/downloads/hugo-extended/hugo-0.87.0/commands/commandeer.go:298 +0x385
github.com/gohugoio/hugo/commands.newCommandeer(0x1, 0xc0001b9180, 0x28e0ee0, 0xc00034c060, 0xc00034c200, 0x0, 0x0, 0x0, 0x2256bc0, 0x1c1b9c01, ...)
	/local/downloads/hugo-extended/hugo-0.87.0/commands/commandeer.go:182 +0x25e
github.com/gohugoio/hugo/commands.initializeConfig(0x2250001, 0xc0001b9180, 0x28e0ee0, 0xc00034c060, 0xc00034c200,0x2b, 0xc0001e7ce0, 0xc000b3fce8)
	/local/downloads/hugo-extended/hugo-0.87.0/commands/hugo.go:123 +0x76
github.com/gohugoio/hugo/commands.(*commandsBuilder).newHugoCmd.func1(0xc000aaf180, 0x37aa840, 0x0, 0x0, 0x0, 0x0)
	/local/downloads/hugo-extended/hugo-0.87.0/commands/commands.go:168 +0x153
github.com/spf13/cobra.(*Command).execute(0xc000aaf180, 0xc0001981d0, 0x0, 0x0, 0xc000aaf180, 0xc0001981d0)
	/home/cgrb/cgrbinst/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:856 +0x472
github.com/spf13/cobra.(*Command).ExecuteC(0xc000aaf180, 0xc00034c060, 0x5, 0x6)
	/home/cgrb/cgrbinst/go/pkg/mod/github.com/spf13/cobra@v1.2.1/command.go:974 +0x375
github.com/gohugoio/hugo/commands.Execute(0xc0001981d0, 0x0, 0x0, 0x4ec245, 0xc00018e058, 0x1ea4ee0, 0x0)
	/local/downloads/hugo-extended/hugo-0.87.0/commands/hugo.go:90 +0xb9
main.main()
	/local/downloads/hugo-extended/hugo-0.87.0/main.go:23 +0x76

$ just deploy-sh
error: Recipe `deploy-sh` could not be run because of an IO error while trying to create a temporary directory or write a file to that directory`:Permission denied (os error 13) at path "/data/justEzdmPq"
```

**Reply by davised** [reply]
Here is when I uncomment the `set export` and `TMPDIR := '/tmp'` lines:

```shell
$ just deploy
git push; dest=/data/www/software/html/updates; hugo; cp -r public/* $dest
Everything up-to-date
Start building sites …
hugo v0.87.0+extended linux/amd64 BuildDate=unknown

                   | EN
-------------------+------
  Pages            | 518
  Paginator pages  |  26
  Non-page files   |  13
  Static files     |  83
  Processed images |   0
  Aliases          | 153
  Sitemaps         |   1
  Cleaned          |   0

Total in 1459 ms
just deploy  9.21s user 1.65s system 359% cpu 3.018 total
```

```shell
$ just deploy-sh
error: Recipe `deploy-sh` could not be run because of an IO error while trying to create a temporary directory or write a file to that directory`:Permission denied (os error 13) at path "/data/justuHC4rd"
```

**Reply by davised** [reply]
If I uncomment `set tmpdir = '/tmp'` alone, I get the hugo errors on both commands (as expected). If I uncomment all three lines, then I can run either recipe.

**Reply by davised** [reply]
So, I think it's all working properly. Since I need to set the env var for hugo, I'll need to have all three lines in my file, but for others who only need a unique tmpdir for just, they will only need the set tmpdir line.

**Reply by casey** [reply]
Awesome, that's great! Thank you for the detailed breakdown 🙏

**Comment by runeimp** [comment]
I think this is a great feature. But, in this case, I think the real fix should be in the cluster management. `TEMPDIR` is such a fundamental thing for the system to get wrong that I'd question the entire setup. If fixing it in Just does work great. But I'd be worried about gremlins scheming to bring down the cluster in general.
